Permissions - Make implied permissions declarative

colemanw opened the pull request:
Overview
----------------------------------------
This improves the implied-permission system to make it simpler for us to say that one permission implies another.

Before
----------------------------------------
Implied permissions were off in their own little space.

After
----------------------------------------
Every permission declaration can include an 'implies' key which tells the permission system which sub-permissions it includes.

colemanw committed:
Permissions - Make implied permissions declarative

Before: Implied permissions were off in their own little space.
After: Every permission declaration can include an 'implies' key which
       tells the permission system which sub-permissions it includes.

eileenmcnaughton commented:
@colemanw I think tests are highlighting some other magic permission you haven't taken into account

colemanw committed:
Permission - Catch runtime exception if permission hook cannot be called

Previously CRM_Core_Permission::check did not rely on calling hook_civicrm_permission and could run at any time.
Now in order to parse implied permissions, we do call the hook and this causes problems with the upgrade tests.
In order to keep compatability with the status-quo, this catches the exception, which means implied permissions
cannot be evaluated at those odd times. OTOH, is there even a user or session yet during those odd times so does it matter?

colemanw commented:
@eileenmcnaughton fixed.

eileenmcnaughton commented:
might be good for @totten to eyeball that last patch

totten commented on CRM/Core/Permission.php:
```diff
@@ -668,6 +694,10 @@ public static function getCorePermissions() {
+        'implies' => [
+          'administer CiviCRM system',
+          'administer CiviCRM data',
+        ],
```
That looks a lot nicer than how it was before. 👍

colemanw commented:
@totten's eyeball didn't seem to have a problem with this PR so I guess that means it's merge ready?

eileenmcnaughton commented:
Yeah - that only needed a light pass by Tim in case of anything that we hadn't thought of
